aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -23,4 +24,128 @@ Tensor& bmm_out_cuda(Tensor &result, const Tensor& batch1, const Tensor& batch2)
   return legacy::cuda::_th_bmm_out(result, batch1, batch2);
 }
 
+Tensor prepare_matrix_for_cublas(Tensor& tensor, bool& transpose_tensor) {
+  Tensor tensor_;
+  IntArrayRef tensor_strides = tensor.strides();
+
+  if ((tensor_strides[0] == 1) && (tensor_strides[1] != 0)) {
+    tensor_ = tensor;
+    transpose_tensor = false;
+  } else if ((tensor_strides[1] == 1) && (tensor_strides[0] != 0)) {
+    tensor_ = tensor;
+    transpose_tensor = true;
+  } else {
+    transpose_tensor = true;
+    tensor_ = tensor.clone(at::MemoryFormat::Contiguous);
+  }
+
+  return tensor_;
+}
+
+// Check https://github.com/pytorch/pytorch/issues/22078
+// for information about the bug. We don't know the exact conditions that trigger it,
+// but using Sgemm or Hgemm on Maxwell or Pascal seems to be a
+// necessary condition.
+static void checkCuda90Bug(int i_m, int i_n, int i_k)
+{
+#if CUDA_VERSION < 9200 && CUDA_VERSION >= 9000
+  static std::once_flag alreadyWarned;
+  const int LIMIT = 1 << 21;
+  if (i_m > LIMIT || i_n > LIMIT || i_k > LIMIT) {
+    cudaDeviceProp* prop = at::cuda::getCurrentDeviceProperties();
+    if (prop->major == 5 || prop->major == 6) {
+      std::call_once(alreadyWarned, []() {
+        TORCH_WARN("Matrix multiplication for dimensions larger than 2^21 has known bugs on your combination of CUDA version and device type. Please consider upgrading to CUDA 9.2 or later.");
+      });
+    }
+  }
+#endif
+}
+
+Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& mat1, const Tensor& mat2, Scalar beta, Scalar alpha) {
+  TORCH_CHECK(
+    (mat1.dim() == 2) && (mat2.dim() == 2) &&
+    (self.dim() == 2) && (result.dim() == 2),
+    "tensors must be 2-D"
+  );
+  IntArrayRef mat1_sizes = mat1.sizes();
+  IntArrayRef mat2_sizes = mat2.sizes();
+  IntArrayRef self_sizes = self.sizes();
+  TORCH_CHECK(mat1_sizes[1] == mat2_sizes[0], "mat1 dim 1 must match mat2 dim 0");
+  TORCH_CHECK(self_sizes[0] == mat1_sizes[0], "self dim 0 must match mat1 dim 0");
+  TORCH_CHECK(self_sizes[1] == mat2_sizes[1], "self dim 1 must match mat2 dim 1");
+
+  // If self and result either point to the same data or if beta is zero,
+  // we can avoid copying self into result. Otherwise, we need to copy.
+  if (result.data_ptr() != self.data_ptr()) {
```

Comment:
do we ensure somewhere that strides of result and self are also the same? It's not important for mm I guess, but for addmm it could be.

Reply:
I wasn't checking that anywhere. I've added it.